# 知識処理特論 最終課題

data: intelliProcess_last.txt (j101-d_11_1499.txt)

In [1]:
import numpy as np
from collections import Counter
from operator import itemgetter

In [9]:
f = open("../data/intelliProcess_last.txt")

text = f.read()
f.close()

## 問１

述べ文字数，および異なり文字数を答えなさい

In [3]:
print("述べ文字数：" + str(len(text)))

diff_char = text[0]
for i in text:
    flag = 0
    for j in diff_char:
        if i == j:
            flag = 1
    if flag == 1:
        continue
    else:
        diff_char += i

print("異なり文字数：" + str(len(diff_char)))

述べ文字数：5960
異なり文字数：450


## 問２

エントロピーを求めることを考える．

(a) 全ての文字種が等確率で出現するとした時の（文字単位の）エントロピー H0 を答えなさい．

(b) 文字の出現頻度（確率）を考慮した場合のエントロピー H1 を答えなさい．

(c) 上記結果を比較し，その考察を記述しなさい．

### (a)

In [4]:
H0 = -np.log2(1/len(diff_char))
H0

8.813781191217037

$$
    H_0 = 8.813781191217037
$$

### (b)

In [5]:
x_list = []
x_list = Counter(text)
H1 = 0

for i in x_list:
    p = x_list[i]/len(text)
    H1 += -p*np.log2(p)

H1

7.459371410966394

$$
    H_1 = 7.459371410966394
$$

### (c)

(a)および(b)より，出現頻度を考慮したエントロピーH_1の方が値が小さい．エントロピーが小さいほど不確実性も低いため，H1の方が優れた結果と言える．

## 問3

(a) 最も単純な方法として，コーパスに出現する全文字種に対してtri-gram統計を求めるとした場合に，必要となる具体的なバイト数を答えなさい．

(b) Nagaoらの手法で必要となる具体的なバイト数を答えなさい．

(c) 上記結果を比較し，その考察を記述しなさい．

### (a)

$$
    ((異なり文字数) \times 2(byte))^3 = (450 \times 2)^3 = 729,000,000 (byte)
$$

### (b)

$$
    7 \times ((述べ文字数) \times 2(byte)) = 7 \times (5960 \times 2) = 83,440 (byte)
$$

### (c)

2つの結果を比較すると，Nagoらの手法の方が圧倒的にバイト数が低いことがわかる．

## 問4

出現頻度の閾値を15とした場合に得られる，6文字以上の the longest compound word を全て答えなさい．

In [6]:
def getNGrams(garm,n):
    return [garm[i:i+n] for i in range(0,len(garm)-(n-1))]

In [7]:
def freqdict(gramf):
    mydict = dict()
    for (ch1,ch2,ch3,ch4,ch5,ch6) in gramf:
        mydict[(ch1,ch2,ch3,ch4,ch5,ch6)] = mydict.get((ch1,ch2,ch3,ch4,ch5,ch6),0)+1
    return mydict

In [8]:
garmf = getNGrams(text,6)
fre = freqdict(garmf)
n_seq = sorted(fre.items(), key=itemgetter(1), reverse=True)
for i in n_seq:
    if i[1] >= 15:
        print(i)

(('ス', 'ラ', 'ッ', 'プ', '奏', '法'), 17)
(('エ', 'レ', 'キ', 'ベ', 'ー', 'ス'), 16)


## 問5

上記で求めた the longest compound word であるwについて，その長さをnとしたとき，n=n1+n2 となるよう，適当にw1,w2に分割し，それぞれの出現頻度を答えなさい．また，この結果について，考察を記述しなさい．

## 問6

受講した感想や質問などを記述しなさい．